scripts/pi-hole/php/groups.php

```diff
@@ -497,18 +497,41 @@ function JSON_error($message = null)
     } catch (\Exception $ex) {
         JSON_error($ex->getMessage());
     }
-} elseif ($_POST['action'] == 'add_domain') {
+} elseif ($_POST['action'] == 'add_domain' || $_POST['action'] == 'replace_domain') {
     // Add new domain
     try {
         $domains = explode(' ', html_entity_decode(trim($_POST['domain'])));
         $before = intval($db->querySingle("SELECT COUNT(*) FROM domainlist;"));
         $total = count($domains);
         $added = 0;
-        $stmt = $db->prepare('REPLACE INTO domainlist (domain,type,comment) VALUES (:domain,:type,:comment)');
-        if (!$stmt) {
+
+        // Prepare INSERT INTO statement
+        $insert_stmt = $db->prepare('REPLACE INTO domainlist (domain,type) VALUES (:domain,:type)');
+        if (!$insert_stmt) {
+            throw new Exception('While preparing statement: ' . $db->lastErrorMsg());
+        }
+
+        // Prepare UPDATE statement
+        $update_stmt = $db->prepare('UPDATE domainlist SET comment = :comment WHERE domain = :domain AND type = :type');
+        if (!$update_stmt) {
             throw new Exception('While preparing statement: ' . $db->lastErrorMsg());
         }
 
+        $check_stmt = null;
+        $delete_stmt = null;
+        if($_POST['action'] == 'replace_domain') {
+            // Check statement will reveal any group associations for a given (domain,type) which do NOT belong to the default group
+            $check_stmt = $db->prepare('SELECT EXISTS(SELECT * FROM domainlist_by_group WHERE domainlist_id = (SELECT id FROM domainlist WHERE domain = :domain) AND group_id != 0)');
```

Comment:
Now that I am awake again, just looking over the `SQL`. Personally not a fan of nested subqueries, can we do a `JOIN` instead please?

```suggestion
            $check_stmt = $db->prepare('SELECT EXISTS(SELECT domain FROM domainlist_by_group dlbg JOIN domainlist dl on dlbg.domainlist_id = dl.id WHERE dl.domain = :domain)');
```

Reply:
(this should also be more perfomant)

Reply:
Who am I to speak up against the godfather of SQL?